---
title: "Final project_Meng Yuan"
author: "Meng Yuan"
date: "11/22/2019"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```


## Data description:
The data I will be using in this final project are publicly available data from the article: 
Wang, Y., Klijn, J. G., Zhang, Y., Sieuwerts, A. M., Look, M. P., Yang, F., â€¦ Foekens, J. A. (2005). Gene-expression profiles to predict distant metastasis of lymph-node-negative primary breast cancer. The Lancet. https://doi.org/10.1016/s0140-6736(05)17947-1. The data were downloaded from NCBI GEO database (GSE2034).

In this study, the researchers examined the expression levels of about 22,000 genes from 286 lymph node negative breast cancer patient samples. Of all 286 patients, 106 had distant metastasis. From the 22,000 genes they generated a 76-gene profile which enabled them to make predictions whether of not the patient would develop distal metastasis.


## To import the patient ID sheet to R (patient IDs were stored in a txt file):

```{r, warning=FALSE, error=FALSE,message=FALSE}
library(readr)
ID <- read.csv("/Users/myuanmoon/GoogleDrive/PhD_studies/Lectures/Lectures_2019Fall/FAES R/Final project/Data to use/GSE2034-22071_ID.txt")
head(ID)
```

## As can be seen above, the ID_REF column contains both GSM numbers and sample indexes with a "=" in between. For the convenience of subsequent analysis, I would like to split the ID_REF column into two, one "GSM" column and one "Sample_index" column:

```{r, warning=FALSE,message=FALSE}
library(tidyverse)
ID_split <- separate(ID, col=ID_REF, into = c("GSM","Sample_index"),sep = "=")
head(ID_split)
```

## Then, I would like to check the data type in ID_split:

The data type for GSM column is:
```{r, warning=FALSE,message=FALSE}
class(ID_split$GSM)
```

The data type for Sample_index column is:
```{r, warning=FALSE,message=FALSE}
class(ID_split$Sample_index)
```

## Let's check if ID_split is a dataframe:
```{r, warning=FALSE,message=FALSE}
is.data.frame(ID_split)
```

## Then it's time to import the gene expression data obtained by the researchers using Affymetrix Human U133a GeneChips (data were stored as xlsx format):

```{r, warning=FALSE, error=FALSE,message=FALSE}
Gene <- readxl::read_excel("/Users/myuanmoon/GoogleDrive/PhD_studies/Lectures/Lectures_2019Fall/FAES R/Final project/Data to use/GSE2034-22071_expression.xlsx")
head(Gene)
```

## Now let's check the data types in Gene:

The data type for ID_REF column is:
```{r, warning=FALSE,message=FALSE}
class(Gene$ID_REF)
```

The data type for other columns are (using the first and the last columns as examples):
```{r, warning=FALSE,message=FALSE}
class(Gene$GSM36777)
class(Gene$GSM37062)
```

## Then let's convert Gene into a dataframe:
```{r, warning=FALSE,message=FALSE}
Gene_dataframe <- as.data.frame(Gene)
class(Gene_dataframe)
```

## The name of the first column in Gene_dataframe is "ID_REF" but this column is actually a list of genes. So I would like to change the column name into "Gene_ID":

```{r, warning=FALSE,message=FALSE}
names(Gene_dataframe)[names(Gene_dataframe)=='ID_REF'] <- 'Gene_ID'
Gene_dataframe[1:10,1:10]
```

## In order to match the table layouts of ID and Gene_dataframe, I would like to gather the sample columns in Gene_dataframe into one column named "GSM" and spread different genes into different columns (basically, after this step the columns in Gene_dataframe will be converted to rows and the rows will become columns):

```{r, warning=FALSE,message=FALSE}
Gene_gather <- gather(Gene_dataframe, key="GSM",value="Expression",-1)
Gene_spread <- spread(Gene_gather, Gene_ID, Expression)
Gene_spread[1:10,1:10]
```

## Now let's combine the ID_split and Gene_spread by GSM column:

```{r, warning=FALSE,message=FALSE}
ID_gene <- left_join(ID_split,Gene_spread,by='GSM')
ID_gene[1:10,1:10]
```

## It looks like the gene expression values won't display in the joined table if the left_join() function is used. Let's try the cbind() function instead:

```{r, warning=FALSE,message=FALSE}
names(Gene_spread)[names(Gene_spread)=='GSM'] <- 'GSM2'
ID_gene2 <- cbind(ID_split,Gene_spread)
ID_gene2[1:10,1:10]
```

## It worked! The GSM columns in ID_split and Gene_spread matched with each other by row in all 286 samples. Next, I need to get rid of the redundant GSM2 column:

```{r, warning=FALSE,message=FALSE}
ID_gene3 <- select(ID_gene2,-3)
ID_gene3[1:10,1:10]
```

## The patient information can be viewed in the GEO database which contains the information whether or not the patient developed distant metastasis. There is no excel or txt file available to be downloaded, so I simply copied and pasted the information into a txt file. Let's import the patient information file now:
```{r, warning=FALSE,message=FALSE}
Patient <- read.csv("/Users/myuanmoon/GoogleDrive/PhD_studies/Lectures/Lectures_2019Fall/FAES R/Final project/Data to use/Patient_info.txt")
head(Patient)
```

## As can be seen above, Patient has only one column which contains all the information. I want to separate each parameter into individual columns:

```{r, warning=FALSE,message=FALSE}
Patient_split <- separate (Patient,col = Info, into = c("ID","GSM","Lymph_node","Time","Relapse","ER","Brain"))
head(Patient_split)
```

## As I will only need the "Relapse" column for subsequent analysis, let's select only the "Relapse" and "GSM" columns and create a new dataset:

```{r, warning=FALSE,message=FALSE}
Patient_select <- select(Patient_split,c("GSM","Relapse"))
head(Patient_select)
```

## The values in "Relapse" column should be factors, let's check if they are factors. If not, I will do the conversion.

```{r, warning=FALSE,message=FALSE}
is.factor(Patient_select$Relapse)
Patient_select$Relapse <- as.factor(Patient_select$Relapse)
class(Patient_select$Relapse)
```

## Now I would like to combine ID_gene3 and Patient_select. But before that, I would need to sort Patient_select by "GSM":

```{r, warning=FALSE,message=FALSE}
Patient_sort <- arrange(Patient_select,GSM)
head(Patient_sort)
names(Patient_sort)[names(Patient_sort)=='GSM'] <- 'GSM2'
ID_gene_patient <- cbind(Patient_sort,ID_gene3)
ID_gene_patient[1:10,1:10]
```

## As we can see, the GSM columns in ID_gene3 and Patient_sort datasets matched by row with each other. Now we can delete the redundent GSM2 column in ID_gene_patient:

```{r, warning=FALSE,message=FALSE}
ID_gene_patient2 <- select(ID_gene_patient,-1)
ID_gene_patient2[1:10,1:10]
```

## Now I would like to divide all patient samples by whether or not the patient had relapse and create two new datasets for that. Analysis will be conducted to compare these two groups:

```{r, warning=FALSE,message=FALSE}
ID_gene_patient_relapse <- filter(ID_gene_patient2,Relapse==1)
ID_gene_patient_relapse[1:10,1:10]
ID_gene_patient_no <- filter(ID_gene_patient2,Relapse==0)
ID_gene_patient_no[1:10,1:10]
```

## In this study the researchers tested about 22,000 genes. I want to use the first 20 genes on the list as example and compare the expression levels of those genes in ID_gene_patient_relapse and ID_gene_patient_no datasets. The aim is to find a group of genes that are differentially expressed in the patients with and without replase:

To do so, I would like to first generate plots of the nine selected genes (relapse group v.s. no relapse group):
```{r, warning=FALSE,message=FALSE}
library(cowplot)
Gene1_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1007_s_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1007_s_at') 
Gene2_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1053_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1053_at')
Gene3_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`117_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '117_at')
Gene4_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`121_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '121_at')
Gene5_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1255_g_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1255_g_at')
Gene6_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1294_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1294_at')
Gene7_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1316_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1316_at')
Gene8_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1320_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1320_at')
Gene9_relapse_no <- ID_gene_patient2 %>% 
  ggplot(aes(x=Relapse,y=`1405_i_at`))+geom_boxplot(mapping=aes(color=Relapse),show.legend = FALSE, )+geom_point(mapping=aes(color=Relapse),show.legend = FALSE) + labs(x='Relapse',y='Expression',title = '1405_i_at')
plot_grid(Gene1_relapse_no,Gene2_relapse_no,Gene3_relapse_no,Gene4_relapse_no,Gene5_relapse_no,Gene6_relapse_no,Gene7_relapse_no,Gene8_relapse_no,Gene9_relapse_no,ncol = 3,labels ="Differential_gene expression_in_patients_with_and_without_relapse") + panel_border()
```

## From the graphs it looks like there is no big difference between the relapse patient group and no relapse group. T-tests are now needed to determine if there are significant differences between the two groups at all:

```{r, warning=FALSE, message=FALSE}
library(broom)
t1077 <- t.test(ID_gene_patient_relapse$`1007_s_at`,ID_gene_patient_no$`1007_s_at`)
ttable1077 <- tidy(t1077)
t1053 <- t.test(ID_gene_patient_relapse$`1053_at`,ID_gene_patient_no$`1053_at`)
ttable1053 <- tidy(t1053)
t117 <- t.test(ID_gene_patient_relapse$`117_at`,ID_gene_patient_no$`117_at`)
ttable117 <- tidy(t117)
t121 <- t.test(ID_gene_patient_relapse$`121_at`,ID_gene_patient_no$`121_at`)
ttable121 <- tidy(t121)
t1255 <- t.test(ID_gene_patient_relapse$`1255_g_at`,ID_gene_patient_no$`1255_g_at`)
ttable1255 <- tidy(t1255)
t1294 <- t.test(ID_gene_patient_relapse$`1294_at`,ID_gene_patient_no$`1294_at`)
ttable1294 <- tidy(t1294)
t1316 <- t.test(ID_gene_patient_relapse$`1316_at`,ID_gene_patient_no$`1316_at`)
ttable1316 <- tidy(t1316)
t1320 <- t.test(ID_gene_patient_relapse$`1320_at`,ID_gene_patient_no$`1320_at`)
ttable1320 <- tidy(t1320)
t1405 <- t.test(ID_gene_patient_relapse$`1405_i_at`,ID_gene_patient_no$`1405_i_at`)
ttable1405 <- tidy(t1405)
tlist <- list(ttable1077,ttable1053,ttable117,ttable121,ttable1255,ttable1294,ttable1316,ttable1320,ttable1405)
tlist
```

## From the p values it could be seen that among the nine selected genes, only 1405_i_at was differentially expressed in patients with and without relapse.