<!DOCTYPE html>
<html lang="" xml:lang="">
  <head>
    <title>Joins, and more plotting</title>
    <meta charset="utf-8" />
    <meta name="author" content="Abhijit Dasgupta" />
    <script src="06-joins-plots_files/header-attrs-2.3/header-attrs.js"></script>
    <link href="06-joins-plots_files/remark-css-0.0.1/default.css" rel="stylesheet" />
    <link rel="stylesheet" href="robot.css" type="text/css" />
    <link rel="stylesheet" href="robot-fonts.css" type="text/css" />
    <link rel="stylesheet" href="sfah.css" type="text/css" />
  </head>
  <body>
    <textarea id="source">
class: center, middle, inverse, title-slide

# Joins, and more plotting
### Abhijit Dasgupta
### Fall, 2019

---



layout: true

&lt;div class="my-header"&gt;
&lt;span&gt;BIOF339, Fall, 2019&lt;/span&gt;&lt;/div&gt;



---

# Goals today 

+ Learn how to join data sets (merging)
+ See how to transform data sets to help our plotting
+ Some additional plot types, and customization

???

We need to 

+ put datasets capturing different attributes together to find a complete picture
+ evaluate different attributes to see if they contribute to our understanding
+ hedge our bets to ensure we find 

---
    
# Data

This data set is taken from a breast cancer proteome database available [here](https://www.kaggle.com/piotrgrabo/breastcancerproteomes) and modified for this exercise.

+ Clinical data: [CSV](data/BreastCancer_Clinical.csv)|[XLSX](data/BreastCancer_Clinical.xlsx)
+ Proteome data: [CSV](data/BreastCancer_Expression.csv)|[XLSX](data/BreastCancer_Expression.xlsx)

---
class: inverse, middle, center

# Joins

---
# Putting data sets together

+ Quite often, data on individuals lie in different tables
    - Clinical, demographic and bioinformatic data
--
    - Drug, procedure, and payment data (think Medicare)
--
    - Personal health data across different healthcare entities

---

# Joining data sets

We already talked about `cbind` and `rbind`:

.pull-left[
&lt;span style="text-align:center;"&gt;`cbind`&lt;/span&gt;

```r
knitr::include_graphics('img/addcol.png')
```

&lt;img src="img/addcol.png" width="640" /&gt;
]
.pull-right[
&lt;span style="text-align:center;"&gt;`rbind`&lt;/span&gt;

```r
knitr::include_graphics('img/addrow.png')
```

&lt;img src="img/addrow.png" width="640" /&gt;
]

---

# Joining data sets

.pull-left[
We will talk about more general ways of joining two datasets

We will assume:

1. We have two rectangular data sets (so `data.frame` or `tibble`)
1. There is at least one variable (column) in common, even if they have different names
    - ID number
    - SSN (Social Security number)
    - Identifiable information
]

.pull-right[
&lt;img src="img/merge.png" height="10%"/&gt;
]

---

# Joining data sets

&lt;img width="100%" src="img/joins.png"/&gt;

--

&lt;table width="100%"&gt;
&lt;tr&gt;
&lt;td style="text-align:center;"&gt;inner_join&lt;/td&gt;
&lt;td style="text-align:center;"&gt;left_join&lt;/td&gt;
&lt;td style="text-align:center;"&gt;right_join&lt;/td&gt;
&lt;td style="text-align:center;"&gt;outer_join&lt;/td&gt;
&lt;/tr&gt;&lt;/table&gt;

--

The "join condition" are the common variables in the two datasets, i.e. rows are selected if the values of the common variables in the left dataset matches the values of the common variables in the right dataset

---

## Data example


```r
library(readxl)
clinical &lt;- read_excel('data/BreastCancer_Clinical.xlsx', 
                       .name_repair='universal') 
proteome &lt;- read_excel('data/BreastCancer_Expression.xlsx', 
                       .name_repair='universal') 
```

.pull-left[

```r
clinical
```

```
#&gt;  # A tibble: 105 x 30
#&gt;     Complete.TCGA.ID Gender Age.at.Initial.… ER.Status PR.Status HER2.Final.Stat…
#&gt;     &lt;chr&gt;            &lt;chr&gt;             &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;           
#&gt;   1 TCGA-A2-A0T2     FEMALE               66 Negative  Negative  Negative        
#&gt;   2 TCGA-A2-A0CM     FEMALE               40 Negative  Negative  Negative        
#&gt;   3 TCGA-BH-A18V     FEMALE               48 Negative  Negative  Negative        
#&gt;   4 TCGA-BH-A18Q     FEMALE               56 Negative  Negative  Negative        
#&gt;   5 TCGA-BH-A0E0     FEMALE               38 Negative  Negative  Negative        
#&gt;   6 TCGA-A7-A0CE     FEMALE               57 Negative  Negative  Negative        
#&gt;   7 TCGA-D8-A142     FEMALE               74 Negative  Negative  Negative        
#&gt;   8 TCGA-A2-A0D0     FEMALE               60 Negative  Negative  Negative        
#&gt;   9 TCGA-AO-A0J6     FEMALE               61 Negative  Negative  Negative        
#&gt;  10 TCGA-A2-A0YM     FEMALE               67 Negative  Negative  Negative        
#&gt;  # … with 95 more rows, and 24 more variables: Tumor &lt;chr&gt;,
#&gt;  #   Tumor..T1.Coded &lt;chr&gt;, Node &lt;chr&gt;, Node.Coded &lt;chr&gt;, Metastasis &lt;chr&gt;,
#&gt;  #   Metastasis.Coded &lt;chr&gt;, AJCC.Stage &lt;chr&gt;, Converted.Stage &lt;chr&gt;,
#&gt;  #   Survival.Data.Form &lt;chr&gt;, Vital.Status &lt;chr&gt;,
#&gt;  #   Days.to.Date.of.Last.Contact &lt;dbl&gt;, Days.to.date.of.Death &lt;dbl&gt;,
#&gt;  #   OS.event &lt;dbl&gt;, OS.Time &lt;dbl&gt;, PAM50.mRNA &lt;chr&gt;,
#&gt;  #   SigClust.Unsupervised.mRNA &lt;dbl&gt;, SigClust.Intrinsic.mRNA &lt;dbl&gt;,
#&gt;  #   miRNA.Clusters &lt;dbl&gt;, methylation.Clusters &lt;dbl&gt;, RPPA.Clusters &lt;chr&gt;,
#&gt;  #   CN.Clusters &lt;dbl&gt;, Integrated.Clusters..with.PAM50. &lt;dbl&gt;,
#&gt;  #   Integrated.Clusters..no.exp. &lt;dbl&gt;, Integrated.Clusters..unsup.exp. &lt;dbl&gt;
```
]
.pull-right[

```r
proteome
```

```
#&gt;  # A tibble: 83 x 11
#&gt;     TCGA_ID NP_958782 NP_958785 NP_958786 NP_000436 NP_958781 NP_958780 NP_958783
#&gt;     &lt;chr&gt;       &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;   1 TCGA-A…     1.10      1.11      1.11      1.11      1.12      1.11      1.11 
#&gt;   2 TCGA-C…     2.61      2.65      2.65      2.65      2.65      2.65      2.65 
#&gt;   3 TCGA-A…    -0.660    -0.649    -0.654    -0.632    -0.640    -0.654    -0.649
#&gt;   4 TCGA-B…     0.195     0.215     0.215     0.205     0.215     0.215     0.215
#&gt;   5 TCGA-C…    -0.494    -0.504    -0.501    -0.510    -0.504    -0.504    -0.501
#&gt;   6 TCGA-C…     2.77      2.78      2.78      2.80      2.79      2.78      2.78 
#&gt;   7 TCGA-E…     0.863     0.870     0.870     0.866     0.870     0.870     0.870
#&gt;   8 TCGA-C…     1.41      1.41      1.41      1.41      1.41      1.41      1.41 
#&gt;   9 TCGA-A…     1.19      1.19      1.19      1.19      1.20      1.19      1.19 
#&gt;  10 TCGA-A…     1.10      1.10      1.10      1.10      1.09      1.10      1.10 
#&gt;  # … with 73 more rows, and 3 more variables: NP_958784 &lt;dbl&gt;, NP_112598 &lt;dbl&gt;,
#&gt;  #   NP_001611 &lt;dbl&gt;
```

]

---

## Data example


```r
library(readxl)
clinical &lt;- read_excel('data/BreastCancer_Clinical.xlsx',
                       .name_repair = 'universal')
proteome &lt;- read_excel('data/BreastCancer_Expression.xlsx',
                         .name_repair = 'universal')
```

.pull-left[

```r
clinical[,1:2]
```

```
#&gt;  # A tibble: 105 x 2
#&gt;     Complete.TCGA.ID Gender
#&gt;     &lt;chr&gt;            &lt;chr&gt; 
#&gt;   1 TCGA-A2-A0T2     FEMALE
#&gt;   2 TCGA-A2-A0CM     FEMALE
#&gt;   3 TCGA-BH-A18V     FEMALE
#&gt;   4 TCGA-BH-A18Q     FEMALE
#&gt;   5 TCGA-BH-A0E0     FEMALE
#&gt;   6 TCGA-A7-A0CE     FEMALE
#&gt;   7 TCGA-D8-A142     FEMALE
#&gt;   8 TCGA-A2-A0D0     FEMALE
#&gt;   9 TCGA-AO-A0J6     FEMALE
#&gt;  10 TCGA-A2-A0YM     FEMALE
#&gt;  # … with 95 more rows
```
]
.pull-right[

```r
proteome[,1:2]
```

```
#&gt;  # A tibble: 83 x 2
#&gt;     TCGA_ID      NP_958782
#&gt;     &lt;chr&gt;            &lt;dbl&gt;
#&gt;   1 TCGA-AO-A12D     1.10 
#&gt;   2 TCGA-C8-A131     2.61 
#&gt;   3 TCGA-AO-A12B    -0.660
#&gt;   4 TCGA-BH-A18Q     0.195
#&gt;   5 TCGA-C8-A130    -0.494
#&gt;   6 TCGA-C8-A138     2.77 
#&gt;   7 TCGA-E2-A154     0.863
#&gt;   8 TCGA-C8-A12L     1.41 
#&gt;   9 TCGA-A2-A0EX     1.19 
#&gt;  10 TCGA-AO-A12D     1.10 
#&gt;  # … with 73 more rows
```
]

--

We see that both have the same ID variable, but with different names and different orders

???

Let's keep only the first two columns so we can see the ID variable

---

## Data example

Let's make sure that the ID's are truly IDs, i.e. each row has a unique value


```r
length(unique(clinical$Complete.TCGA.ID)) == nrow(clinical)
```

```
#&gt;  [1] TRUE
```
--


```r
length(unique(proteome$TCGA_ID)) == nrow(proteome)
```

```
#&gt;  [1] FALSE
```


--
&lt;div style="height:25%;margins:auto;"&gt;
&lt;img style="display:block; margin:0 auto; height: 70%;" src="https://twitchy.com/wp-content/uploads/2015/04/screen-shot-2015-04-13-at-2-06-38-pm-300x300.png"/&gt;
&lt;/div&gt;

???

We need the ID variables to be unique for each row. If we use multiple columns to define the "ID" then each row needs to have a unique set of values for those columns. Otherwise the joins get confused about 
which rows go with which rows. 

---
## Data example

For convenience we'll keep the first instance for each ID in the `proteome` data


```r
proteome &lt;- proteome %&gt;% filter(!duplicated(TCGA_ID))
```

&gt; `duplicated` = TRUE if a previous row contains the same value

--


```r
length(unique(proteome$TCGA_ID)) == nrow(proteome)
```

```
#&gt;  [1] TRUE
```

???

We don't have to sort data for duplicated

---
## Inner join


```r
common_rows &lt;- inner_join(clinical[,1:6], proteome, by=c('Complete.TCGA.ID'='TCGA_ID'))
```

```
#&gt;  # A tibble: 77 x 16
#&gt;     Complete.TCGA.ID Gender Age.at.Initial.… ER.Status PR.Status HER2.Final.Stat…
#&gt;     &lt;chr&gt;            &lt;chr&gt;             &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;     &lt;chr&gt;           
#&gt;   1 TCGA-A2-A0CM     FEMALE               40 Negative  Negative  Negative        
#&gt;   2 TCGA-BH-A18Q     FEMALE               56 Negative  Negative  Negative        
#&gt;   3 TCGA-A7-A0CE     FEMALE               57 Negative  Negative  Negative        
#&gt;   4 TCGA-D8-A142     FEMALE               74 Negative  Negative  Negative        
#&gt;   5 TCGA-AO-A0J6     FEMALE               61 Negative  Negative  Negative        
#&gt;   6 TCGA-A2-A0YM     FEMALE               67 Negative  Negative  Negative        
#&gt;   7 TCGA-A2-A0D2     FEMALE               45 Negative  Negative  Negative        
#&gt;   8 TCGA-A2-A0SX     FEMALE               48 Negative  Negative  Negative        
#&gt;   9 TCGA-AO-A0JL     FEMALE               59 Negative  Negative  Negative        
#&gt;  10 TCGA-AO-A12F     FEMALE               36 Negative  Negative  Negative        
#&gt;  # … with 67 more rows, and 10 more variables: NP_958782 &lt;dbl&gt;, NP_958785 &lt;dbl&gt;,
#&gt;  #   NP_958786 &lt;dbl&gt;, NP_000436 &lt;dbl&gt;, NP_958781 &lt;dbl&gt;, NP_958780 &lt;dbl&gt;,
#&gt;  #   NP_958783 &lt;dbl&gt;, NP_958784 &lt;dbl&gt;, NP_112598 &lt;dbl&gt;, NP_001611 &lt;dbl&gt;
```

--

Note that we have all the columns from both datasets, but only 77 rows, which is the common set of IDs from the two datasets
--

&gt; If you don't include the `by` option, R will attempt to match values of any columns with the same names

---
## Left join

```r
left_rows &lt;- left_join(clinical[,1:6], proteome, by=c('Complete.TCGA.ID'='TCGA_ID'))
```

```
#&gt;  # A tibble: 105 x 16
#&gt;    Complete.TCGA.ID Gender Age.at.Initial.Pathologic.Diagnosis ER.Status
#&gt;    &lt;chr&gt;            &lt;chr&gt;                                &lt;dbl&gt; &lt;chr&gt;    
#&gt;  1 TCGA-A2-A0T2     FEMALE                                  66 Negative 
#&gt;  2 TCGA-A2-A0CM     FEMALE                                  40 Negative 
#&gt;  3 TCGA-BH-A18V     FEMALE                                  48 Negative 
#&gt;    PR.Status HER2.Final.Status NP_958782 NP_958785 NP_958786 NP_000436 NP_958781
#&gt;    &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1 Negative  Negative             NA        NA        NA        NA        NA    
#&gt;  2 Negative  Negative              0.683     0.694     0.698     0.687     0.687
#&gt;  3 Negative  Negative             NA        NA        NA        NA        NA    
#&gt;    NP_958780 NP_958783 NP_958784 NP_112598 NP_001611
#&gt;        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1    NA        NA        NA         NA       NA    
#&gt;  2     0.698     0.698     0.698     -2.65    -0.984
#&gt;  3    NA        NA        NA         NA       NA    
#&gt;  # … with 102 more rows
```

We get 105 rows, which is all the rows of `clinical`, combined with the rows of `proteome` with common IDs. The rest of the rows get `NA` for the proteome columns.

---
## Right join

```r
right_rows &lt;- right_join(clinical[,1:6], proteome, by=c('Complete.TCGA.ID'='TCGA_ID'))
```

```
#&gt;  # A tibble: 80 x 16
#&gt;    Complete.TCGA.ID Gender Age.at.Initial.Pathologic.Diagnosis ER.Status
#&gt;    &lt;chr&gt;            &lt;chr&gt;                                &lt;dbl&gt; &lt;chr&gt;    
#&gt;  1 TCGA-A2-A0CM     FEMALE                                  40 Negative 
#&gt;  2 TCGA-BH-A18Q     FEMALE                                  56 Negative 
#&gt;  3 TCGA-A7-A0CE     FEMALE                                  57 Negative 
#&gt;    PR.Status HER2.Final.Status NP_958782 NP_958785 NP_958786 NP_000436 NP_958781
#&gt;    &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1 Negative  Negative              0.683     0.694     0.698     0.687     0.687
#&gt;  2 Negative  Negative              0.195     0.215     0.215     0.205     0.215
#&gt;  3 Negative  Negative             -1.12     -1.12     -1.12     -1.13     -1.13 
#&gt;    NP_958780 NP_958783 NP_958784 NP_112598 NP_001611
#&gt;        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1     0.698     0.698     0.698     -2.65    -0.984
#&gt;  2     0.215     0.215     0.215     -1.04    -0.517
#&gt;  3    -1.12     -1.12     -1.12       2.24    -2.58 
#&gt;  # … with 77 more rows
```

--

Here we get 80 rows, which is all the rows of `proteome`, along with the rows of `clinical` with common IDs, but with the columns of `clinical` appearing first.

---
## Outer/Full Join


```r
full_rows &lt;- full_join(clinical[,1:6], proteome, by=c('Complete.TCGA.ID'='TCGA_ID'))
```

```
#&gt;  # A tibble: 108 x 16
#&gt;    Complete.TCGA.ID Gender Age.at.Initial.Pathologic.Diagnosis ER.Status
#&gt;    &lt;chr&gt;            &lt;chr&gt;                                &lt;dbl&gt; &lt;chr&gt;    
#&gt;  1 TCGA-A2-A0T2     FEMALE                                  66 Negative 
#&gt;  2 TCGA-A2-A0CM     FEMALE                                  40 Negative 
#&gt;  3 TCGA-BH-A18V     FEMALE                                  48 Negative 
#&gt;    PR.Status HER2.Final.Status NP_958782 NP_958785 NP_958786 NP_000436 NP_958781
#&gt;    &lt;chr&gt;     &lt;chr&gt;                 &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1 Negative  Negative             NA        NA        NA        NA        NA    
#&gt;  2 Negative  Negative              0.683     0.694     0.698     0.687     0.687
#&gt;  3 Negative  Negative             NA        NA        NA        NA        NA    
#&gt;    NP_958780 NP_958783 NP_958784 NP_112598 NP_001611
#&gt;        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1    NA        NA        NA         NA       NA    
#&gt;  2     0.698     0.698     0.698     -2.65    -0.984
#&gt;  3    NA        NA        NA         NA       NA    
#&gt;  # … with 105 more rows
```

--

Here we obtain 108 rows and 16 columns. So we've expanded the data in both rows and columns, putting missing values in where needed.

---
# Joins

In each of `inner_join`, `left_join`, `right_join` and `full_join`, the number of columns always increases 

There are also two joins where the number of columns don't increase. They aren't really "joins" in that sense, but really fancy filters on a dataset


```r
tbl &lt;- tribble(~Join,~Use,~Description,
               "semi_join", "semi_join(A,B)", "Keep rows in A where ID matches some ID value in B",
               'anti_join', 'anti_join(A,B)', 'Keep rows in A where ID does NOT match any ID value in B')
knitr::kable(tbl, format='html')
```

&lt;table&gt;
 &lt;thead&gt;
  &lt;tr&gt;
   &lt;th style="text-align:left;"&gt; Join &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Use &lt;/th&gt;
   &lt;th style="text-align:left;"&gt; Description &lt;/th&gt;
  &lt;/tr&gt;
 &lt;/thead&gt;
&lt;tbody&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; semi_join &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; semi_join(A,B) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Keep rows in A where ID matches some ID value in B &lt;/td&gt;
  &lt;/tr&gt;
  &lt;tr&gt;
   &lt;td style="text-align:left;"&gt; anti_join &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; anti_join(A,B) &lt;/td&gt;
   &lt;td style="text-align:left;"&gt; Keep rows in A where ID does NOT match any ID value in B &lt;/td&gt;
  &lt;/tr&gt;
&lt;/tbody&gt;
&lt;/table&gt;

These just filter the rows of `A` without adding any columns of `B`.

---
class: inverse, middle, center

## Are there protein expression differences between ER +ve and ER -ve breast cancers

---

# Create analytic dataset


```r
final_data &lt;- clinical %&gt;% 
  inner_join(proteome, by=c("Complete.TCGA.ID"="TCGA_ID")) %&gt;% 
  filter(Gender =='FEMALE') %&gt;% 
  select(Complete.TCGA.ID, Age.at.Initial.Pathologic.Diagnosis, ER.Status,
         starts_with("NP")) # grabs all the protein data
```

```
#&gt;  # A tibble: 75 x 13
#&gt;    Complete.TCGA.ID Age.at.Initial.Pathologic.Diagnosis ER.Status NP_958782
#&gt;    &lt;chr&gt;                                          &lt;dbl&gt; &lt;chr&gt;         &lt;dbl&gt;
#&gt;  1 TCGA-A2-A0CM                                      40 Negative      0.683
#&gt;  2 TCGA-BH-A18Q                                      56 Negative      0.195
#&gt;  3 TCGA-A7-A0CE                                      57 Negative     -1.12 
#&gt;    NP_958785 NP_958786 NP_000436 NP_958781 NP_958780 NP_958783 NP_958784
#&gt;        &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1     0.694     0.698     0.687     0.687     0.698     0.698     0.698
#&gt;  2     0.215     0.215     0.205     0.215     0.215     0.215     0.215
#&gt;  3    -1.12     -1.12     -1.13     -1.13     -1.12     -1.12     -1.12 
#&gt;    NP_112598 NP_001611
#&gt;        &lt;dbl&gt;     &lt;dbl&gt;
#&gt;  1     -2.65    -0.984
#&gt;  2     -1.04    -0.517
#&gt;  3      2.24    -2.58 
#&gt;  # … with 72 more rows
```

---
## Protein-specific graphs

We want to graph each protein separately, while maintaining alignment with ER status and age.

--

The R trick is to make this wide table long, so you can split on the rows 


```r
final_data2 &lt;- final_data %&gt;% tidyr::gather(protein, expression, starts_with('NP')) %&gt;% 
  arrange(Complete.TCGA.ID)
```

```
#&gt;  # A tibble: 750 x 5
#&gt;    Complete.TCGA.ID Age.at.Initial.Pathologic.Diagnosis ER.Status protein  
#&gt;    &lt;chr&gt;                                          &lt;dbl&gt; &lt;chr&gt;     &lt;chr&gt;    
#&gt;  1 TCGA-A2-A0CM                                      40 Negative  NP_958782
#&gt;  2 TCGA-A2-A0CM                                      40 Negative  NP_958785
#&gt;  3 TCGA-A2-A0CM                                      40 Negative  NP_958786
#&gt;    expression
#&gt;         &lt;dbl&gt;
#&gt;  1      0.683
#&gt;  2      0.694
#&gt;  3      0.698
#&gt;  # … with 747 more rows
```

---
## Facetted plots (Trellis graphics)

Realize that the `group` or `color` or `fill` or similar modifications of geoms are really 
splitting the data based on the grouping variable and then plotting the data from each of the divided datasets

--

&lt;h1 style='text-align:center;'&gt;Split-apply-combine&lt;/h1&gt;

---

## Facetted plots (Trellis graphics)

.pull-left[

```r
ggplot(final_data2, aes(x = expression)) + 
  geom_histogram() +
  facet_wrap(~protein)
```

Here we're splitting the **rows** of the data based on the value of `protein` (which are the protein names), and the plotting a histogram of expression for each subgroup, and then 
putting all the plots back together
]

.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-26-1.png)&lt;!-- --&gt;
]

---
## Facetted plots (Trellis graphics)

.pull-left[

```r
p &lt;- ggplot(final_data2, 
       aes(x = ER.Status, y = expression, 
           color = ER.Status, 
           shape=ER.Status)) + 
  geom_boxplot() + 
  geom_jitter() +
  facet_wrap(~protein)

p
```
]
.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-27-1.png)&lt;!-- --&gt;
]

???

Harsh colors

---
## Facetted plots (Trellis graphics)

.pull-left[

```r
p &lt;- ggplot(final_data2, 
       aes(x = ER.Status, y = expression, 
           color = ER.Status, 
           shape=ER.Status)) + 
  geom_boxplot() + 
  geom_jitter()+
  facet_wrap(~protein) +
* theme_bw() +
* scale_color_manual(values = c("#00AFBB", "#E7B800"))

p
```
]
.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-28-1.png)&lt;!-- --&gt;
]

---
## Facetted plots (Trellis graphics)

.pull-left[

```r
p &lt;- ggplot(final_data2, 
       aes(x = ER.Status, y = expression, 
           color = ER.Status, 
           shape=ER.Status)) + 
  geom_boxplot() + 
  geom_jitter()+
  facet_wrap(~protein) +
  theme_bw() + 
  scale_color_manual(values = c("#00AFBB", "#E7B800")) +
* theme(legend.position='top')

p
```
]
.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-29-1.png)&lt;!-- --&gt;
]

---
## Facetted plots (Trellis graphics)

.pull-left[

```r
p &lt;- ggplot(final_data2, 
       aes(x = ER.Status, y = expression, 
           color = ER.Status, 
           shape=ER.Status)) + 
  geom_boxplot() + 
  geom_jitter()+ 
  facet_wrap(~protein) +
  scale_color_manual(values = c("#00AFBB", "#E7B800")) + 
  theme_bw() + 
  theme(legend.position='top') 
*p + ggpubr::stat_compare_means(size = 2, na.rm=T)
```

Adding statistics

The statistics are computed on each subgroup, so proving that this is really an example of split-apply-combine
]
.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-30-1.png)&lt;!-- --&gt;
]

---
class: middle, center, inverse

&gt; Design is choice. The theory of the visual display of quantitative information consists of principles that generate design options and that guide choices among options. The principles should not be applied rigidly or in a peevish spirit; they are not logically or mathematically certain; and it is better to violate any principle than to place graceless or inelegant marks on paper. Most principles of design should be greeted with some skepticism, for word authority can dominate our vision, and we may come to see only through the lenses of word authority rather than with our own eyes.
&gt;
&gt; --- Edward Tufte, &lt;cite&gt;The Visual Display of Quantitative Data&lt;/cite&gt;
---
class: inverse, middle, center

# Considerations

---

## Tufte's Principles of Graphical Integrity

1. Show data variation, not design variation
1. Do not use graphics to quote data out of context
1. Use clear, detailed, thorough labelling. 
1. Representation of numbers should be directly proportional to numerical quantities
1. Don't use more dimensions than the data require

---

## Tufte's Principles of Graphical Integrity

1. Show data variation, not design variation
    - Don't get fancy, let the data speak
1. Do not use graphics to quote data out of context
    - Maintain accuracy
1. Use clear, detailed, thorough labelling. 
    - Use annotations to make your point
1. Representation of numbers should be directly proportional to numerical quantities
    - This is essential for fair representation
1. Don't use more dimensions than the data require
    - Be appropriate in use of 3D graphics, for example

---
## Tufte's Fundamental Principles of Design

1. Show comparisons
1. Show causality
1. Use multivariate data
1. Completely integrate modes (like text, images, numbers)
1. Establish credibility
1. Focus on content

---

## 7 Basic Rules for Making Charts and Graphs

1. Check the data
1. Explain encodings
1. Label axes
1. Include units
1. Keep your geometry in check
1. Include your sources
1. Consider your audience

&lt;div style='text-align: right'&gt;
Nathan Yau, Flowing Data&lt;br&gt;
&lt;a href="https://flowingdata.com/2010/07/22/7-basic-rules-for-making-charts-and-graphs/"&gt;https://flowingdata.com/2010/07/22/7-basic-rules-for-making-charts-and-graphs/&lt;/a&gt;
&lt;/div&gt;

---

## Presentations and formal papers

.pull-left[
Work on formating

- Fonts
- Colors
- Glyphs
- Labeling
- Panels/Facets and organization
]
.pull-right[
Check any particular requirements from publisher
- Resolution
- File type
- Typically TIFF at 300dpi is required
]

---

## A note on TIFFs

- R creates graphs at 72dpi by default
- I've had most success creating PDFs or SVGs and converting them
- Adobe Acrobat Pro will save PDFs to TIFFs, as will Adobe Illustrator for SVGs
- Make sure you use LZW compression, otherwise you'll fail file size requirements

#### Using [Ghostscript](https://www.ghostscript.com/)
```
gs -q -dNOPAUSE -dBATCH -sDEVICE=tiff24nc -sCompression=lzw -r300x300 -sOutputFile=&lt;output file&gt; &lt;input file&gt;
```
On Windows, replace `gs` with `gswin32c`

#### Using ggplot (appears to give right DPI, but doesn't seem to compress, so files are too big)
```
ggsave('out.tiff', units='in', width=4, height=4, compression = 'lzw', dpi = 300)
```
---


## File Formats

In general, you will generate a graphics file for your plot by calling a function which will have the same name as the desired file format (svg, pdf, jpeg, etc).


```r
library(ggplot2, quietly = TRUE)
svg(filename="myPlot.svg", width = 3, height=3, pointsize = 8)
ggplot(cars, aes(x=speed)) + geom_density()
dev.off()
```
The second command opens a file for output, the third generates the plot, and the fourth command (dev.off()) finishes writing the file and closes it.  By default, graphics go to the last graphics "device" you created and dev.off closes the last graphics device created.

--

A shortcut for this in `ggplot2` is


```r
ggplot(cars, aes(x = speed)) + geom_density()
ggsave('myPlot.svg') # Type is picked up from last 3 letters after .
```


---

## Vector versus Raster (pixel) graphics


```r
pdf(file = "test.pdf", width=3, height=3)
ggplot(cars, aes(x=speed)) + geom_density()
dev.off()
png(filename = "test.png", width=3, height=3, units = "in",res = 100)
ggplot(cars, aes(x=speed)) + geom_density()
dev.off()
```

--

## File Size

test.pdf: 9KB

test.png: 16KB

Raster graphics take more space but give worse results!  In general, you will be better off using vector graphcics when makeing plots and graphs.

---

## Error Bars

You create error bars in ggplot by adding an extra plot argument, geom_errorbar.  You specify the top and bottom "y" position of the error bar, and optionally the width.  You will have to calculate where the error bars should be and choose what they should represent (standard deviation, standard error, 95% confidence interval).

---

## Calculating Standard Error

A standard method to achieve error bars would be to calculate standard error  and store the value in a column.  
Be careful!  There is a standard error function in R (stderr) that has nothing to do with standard errors!  But you can define your on function to calculate it or use a package that supplies the standard error function. 
 

```r
sem &lt;- function(x) sqrt(var(x, na.rm=T)/sum(!is.na(x)))
cars %&gt;% group_by(speed) %&gt;% 
  summarize(meanDist = mean(dist),
            semDist = sem(dist))
```

```
#&gt;  # A tibble: 19 x 3
#&gt;     speed meanDist semDist
#&gt;     &lt;dbl&gt;    &lt;dbl&gt;   &lt;dbl&gt;
#&gt;   1     4      6      4   
#&gt;   2     7     13      9   
#&gt;   3     8     16     NA   
#&gt;   4     9     10     NA   
#&gt;   5    10     26      4.62
#&gt;   6    11     22.5    5.5 
#&gt;   7    12     21.5    2.99
#&gt;   8    13     35      4.12
#&gt;   9    14     50.5   12.1 
#&gt;  10    15     33.3   10.5 
#&gt;  11    16     36      4   
#&gt;  12    17     40.7    5.21
#&gt;  13    18     64.5    9.54
#&gt;  14    19     50      9.45
#&gt;  15    20     50.4    5.31
#&gt;  16    22     66     NA   
#&gt;  17    23     54     NA   
#&gt;  18    24     93.8   10.2 
#&gt;  19    25     85     NA
```

---

## Error Bar Example


```r
sem &lt;- function(x) sqrt(var(x, na.rm=T)/sum(!is.na(x)))
cars %&gt;% group_by(speed) %&gt;% summarize(meanDist = mean(dist),
                                       semDist = sem(dist)) %&gt;% 
ggplot(aes(x=speed, y=meanDist)) + geom_bar(stat="identity") +
  geom_errorbar(aes(ymin=meanDist - semDist, ymax= meanDist+semDist))
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-35-1.png)&lt;!-- --&gt;

---


## Changing the axis title


```r
ggplot(chickwts, aes(x=feed, y=weight)) + geom_boxplot() + 
  labs(x = 'Feed Type', y = 'Chick Weight')
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-36-1.png)&lt;!-- --&gt;

---

## Changing the axis limits


```r
ggplot(chickwts, aes(x=feed, y=weight)) + geom_boxplot() + 
  scale_y_continuous("Chick Weight", limits=c(0,500)) +
  scale_x_discrete("Feed Type")
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-37-1.png)&lt;!-- --&gt;

---

## Log scale


```r
ggplot(chickwts, aes(x=feed, y=weight)) + geom_boxplot() + 
  scale_y_log10("Chick Weight", limits=c(10,500)) +
  scale_x_discrete("Feed Type")
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-38-1.png)&lt;!-- --&gt;

---

## Coloring of Factors


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500))+
  labs(fill = 'Feed')
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-39-1.png)&lt;!-- --&gt;

---

## Manual Color Scale


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500)) +
  scale_fill_manual("Feed Type",values = c("red","orange","yellow","green","blue","violet"))
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-40-1.png)&lt;!-- --&gt;

---

## Fill Color Brewer


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500)) +
  scale_fill_brewer("Feed Type")
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-41-1.png)&lt;!-- --&gt;

---

## Fill Color Brewer


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500)) +
  scale_fill_grey("Feed Type")
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-42-1.png)&lt;!-- --&gt;

---

## Changing Plot "Theme"


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500)) +
  scale_fill_discrete("Feed Type") + theme_bw()
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-43-1.png)&lt;!-- --&gt;

---

## Minimal Theme


```r
ggplot(chickwts, aes(x=weight, fill=feed)) + geom_density(alpha=0.5) + 
  scale_x_continuous("Chick Weight", limits=c(0,500)) +
  scale_fill_discrete("Feed Type") + theme_minimal()
```

![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-44-1.png)&lt;!-- --&gt;

---

## Personal theme

.pull-left[

```r
mytheme &lt;- theme(axis.ticks = element_blank(),
                 axis.text = element_text(color = 'blue', family='Times New Roman'),
                 axis.title = element_text(color = 'red', family = 'Times New Roman',
                                           size = 20))
ggplot(chickwts, aes(x = weight, fill = feed))+geom_density(alpha = 0.5) +
  labs(x = 'Chick Weight', y = '', fill  = 'Feed Type') + 
  mytheme
```
]
.pull-right[
![](/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-46-1.png)&lt;!-- --&gt;
]

---

## Multiple Graphs

The packages `cowplot` and [`ggpubr`](https://rpkgs.datanovia.com/ggpubr/index.html) make putting different graphs on the same panel pretty straightforward.


```r
# install.packages('cowplot')
library(cowplot)
p1 &lt;- ggplot(iris, aes(Sepal.Length, Sepal.Width, color = Species)) + 
  geom_point() + facet_grid(. ~ Species) + stat_smooth(method = "lm") +
  background_grid(major = 'y', minor = "none") + 
  panel_border() + theme(legend.position = "none")

# plot B
p2 &lt;- ggplot(iris, aes(Sepal.Length, fill = Species)) +
  geom_density(alpha = .7) + theme(legend.justification = "top")
p2a &lt;- p2 + theme(legend.position = "none")

# plot C
p3 &lt;- ggplot(iris, aes(Sepal.Width, fill = Species)) +
  geom_density(alpha = .7) + theme(legend.position = "none")

# legend
legend &lt;- get_legend(p2)

# align all plots vertically
plots &lt;- align_plots(p1, p2a, p3, align = 'v', axis = 'l')

# put together bottom row and then everything
bottom_row &lt;- plot_grid(plots[[2]], plots[[3]], legend, labels = c("B", "C"), rel_widths = c(1, 1, .3), nrow = 1)
plot_grid(plots[[1]], bottom_row, labels = c("A"), ncol = 1)
```

---

## Multiple Graphs

&lt;img src="/Users/abhijit/ARAASTAT/Teaching/BIOF339/docs/slides/lectures/06-joins-plots_files/figure-html/06-joins-plots-48-1.png" style="display: block; margin: auto;" /&gt;

---

---

## Fine-tuning the theme


```r
plt &lt;- ggplot(bl, aes(x = estimate)) + geom_histogram(bins = 50)+#geom_density() +
  facet_grid(Event ~ Race, scales = 'free', switch = 'y', space = 'free_x') +
  geom_vline(xintercept = 1, linetype = 2) +
 geom_segment(data = bl2, aes(x = estimate, xend=estimate, yend = 5, y = hts),
               color='red', size = 1.5, arrow = arrow(length = unit(.2, 'cm')))+
  scale_x_continuous(breaks = c(1, seq(0.7, 1.8, by = 0.2)))+ # Unified the x-axis ticks
  labs(x = 'Adjusted HR, compared to Whites', y = '') +
* theme(strip.text = element_text(size = 14, face = 'bold'),
*       strip.text.y = element_text(angle = 180), # Rotate the y-axis labels
*       strip.background.x = element_rect(fill = 'white'),
*       strip.placement = 'outside', # Move labels outside the borders
*       axis.text.y = element_blank(),
*       axis.ticks.y = element_blank(),
*       axis.text.x = element_text(size = 8),
*       panel.spacing.x = unit(2, 'lines'))
```

---

&lt;img src="img/Figure2a.png" width="2400" /&gt;

--- 

# Exporting to PowerPoint

The `export` package allows exporting ggplot graphs to PowerPoint as **editable images**



```r
export::graph2ppt(plt, file="Plots.ppt")
```

    </textarea>
<style data-target="print-only">@media screen {.remark-slide-container{display:block;}.remark-slide-scaler{box-shadow:none;}}</style>
<script src="https://remarkjs.com/downloads/remark-latest.min.js"></script>
<script>var slideshow = remark.create({
"ratio": "16:9",
"highlightLanguage": "r",
"countIncrementalSlides": false,
"highlightStyle": "tomorrow-night-bright",
"highlightLines": true,
"slideNumberFormat": "%current%"
});
if (window.HTMLWidgets) slideshow.on('afterShowSlide', function (slide) {
  window.dispatchEvent(new Event('resize'));
});
(function(d) {
  var s = d.createElement("style"), r = d.querySelector(".remark-slide-scaler");
  if (!r) return;
  s.type = "text/css"; s.innerHTML = "@page {size: " + r.style.width + " " + r.style.height +"; }";
  d.head.appendChild(s);
})(document);

(function(d) {
  var el = d.getElementsByClassName("remark-slides-area");
  if (!el) return;
  var slide, slides = slideshow.getSlides(), els = el[0].children;
  for (var i = 1; i < slides.length; i++) {
    slide = slides[i];
    if (slide.properties.continued === "true" || slide.properties.count === "false") {
      els[i - 1].className += ' has-continuation';
    }
  }
  var s = d.createElement("style");
  s.type = "text/css"; s.innerHTML = "@media print { .has-continuation { display: none; } }";
  d.head.appendChild(s);
})(document);
// delete the temporary CSS (for displaying all slides initially) when the user
// starts to view slides
(function() {
  var deleted = false;
  slideshow.on('beforeShowSlide', function(slide) {
    if (deleted) return;
    var sheets = document.styleSheets, node;
    for (var i = 0; i < sheets.length; i++) {
      node = sheets[i].ownerNode;
      if (node.dataset["target"] !== "print-only") continue;
      node.parentNode.removeChild(node);
    }
    deleted = true;
  });
})();
(function() {
  "use strict"
  // Replace <script> tags in slides area to make them executable
  var scripts = document.querySelectorAll(
    '.remark-slides-area .remark-slide-container script'
  );
  if (!scripts.length) return;
  for (var i = 0; i < scripts.length; i++) {
    var s = document.createElement('script');
    var code = document.createTextNode(scripts[i].textContent);
    s.appendChild(code);
    var scriptAttrs = scripts[i].attributes;
    for (var j = 0; j < scriptAttrs.length; j++) {
      s.setAttribute(scriptAttrs[j].name, scriptAttrs[j].value);
    }
    scripts[i].parentElement.replaceChild(s, scripts[i]);
  }
})();
(function() {
  var links = document.getElementsByTagName('a');
  for (var i = 0; i < links.length; i++) {
    if (/^(https?:)?\/\//.test(links[i].getAttribute('href'))) {
      links[i].target = '_blank';
    }
  }
})();
// adds .remark-code-has-line-highlighted class to <pre> parent elements
// of code chunks containing highlighted lines with class .remark-code-line-highlighted
(function(d) {
  const hlines = d.querySelectorAll('.remark-code-line-highlighted');
  const preParents = [];
  const findPreParent = function(line, p = 0) {
    if (p > 1) return null; // traverse up no further than grandparent
    const el = line.parentElement;
    return el.tagName === "PRE" ? el : findPreParent(el, ++p);
  };

  for (let line of hlines) {
    let pre = findPreParent(line);
    if (pre && !preParents.includes(pre)) preParents.push(pre);
  }
  preParents.forEach(p => p.classList.add("remark-code-has-line-highlighted"));
})(document);</script>

<script>
slideshow._releaseMath = function(el) {
  var i, text, code, codes = el.getElementsByTagName('code');
  for (i = 0; i < codes.length;) {
    code = codes[i];
    if (code.parentNode.tagName !== 'PRE' && code.childElementCount === 0) {
      text = code.textContent;
      if (/^\\\((.|\s)+\\\)$/.test(text) || /^\\\[(.|\s)+\\\]$/.test(text) ||
          /^\$\$(.|\s)+\$\$$/.test(text) ||
          /^\\begin\{([^}]+)\}(.|\s)+\\end\{[^}]+\}$/.test(text)) {
        code.outerHTML = code.innerHTML;  // remove <code></code>
        continue;
      }
    }
    i++;
  }
};
slideshow._releaseMath(document);
</script>
<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
(function () {
  var script = document.createElement('script');
  script.type = 'text/javascript';
  script.src  = 'https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML';
  if (location.protocol !== 'file:' && /^https?:/.test(script.src))
    script.src  = script.src.replace(/^https?:/, '');
  document.getElementsByTagName('head')[0].appendChild(script);
})();
</script>
  </body>
</html>
