---
title: "Final_Alexander_Katherine"
author: "Katherine Alexander"
date: "12/13/2019"
output: 
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
#Load necessary packages
library(readr)
library(dplyr)
library(tidyverse)
library(ggplot2)
library(knitr)
library(readr)
library(rio)
knitr::opts_chunk$set(message=F, warning=F)
#Read data into R
metabolism.genes <- rio::import("/Users/katherinealexander/Documents/Metabolism.Genes.Expt_339.csv")
# View(metabolism.genes) 
head(metabolism.genes)
```

## Introduction
We are interested in the relationship between Adenosine A2AR and metabolism gene expression levels in healthy vs. diseased states. The genes under study code for crucial enzymes involved in metabolism and will be correlated to the expression of the gene of interest, denoted as Adora2a, which codes for the adenosine A2A receptor. This target has been implicated in many pathophysiologies, from Parkinson's Disease to depression to cancer, and is associated with immunosuppression within microenvironments of varying cell types. By understanding the correlation of metabolism-associated gene expression with Adora2a, and its other family members (Adora1, Adora2b, Adora3), we can seek insights into its impact on metabolic flux.

```{r pressure, echo=FALSE, fig.cap="Heat map demonstrating correlation of metabolism gene expression levels to that of Adora2a.", out.width = '30%'}
knitr::include_graphics("/Users/katherinealexander/Desktop/Screen Shot 2019-12-18 at 4.58.33 PM.png")
```

# Tidying and Data Munging

We are using the dplyr package to remove the "Pathways" column, as well as an empty column included in the dataset.
```{r}
# Drop the columns of the dataframe
metabolism.genes_clean <- metabolism.genes %>% 
  select (-c(10,11)) %>% 
  gather(Samples, Expression, 2:9) %>% 
  separate(Samples, into = c("Health.Status", "Sample.Number"), sep = ".No.Tx.") %>% 
  arrange(Gene.Symbol)
```

## Statistical Analyses


```{r}
# library(broom)

t.test(`Expression`~Health.Status, data=metabolism.genes_clean)

wilcox.test(`Expression`~Health.Status, data=metabolism.genes_clean)

aov(`Expression`~Health.Status, data=metabolism.genes_clean)

kruskal.test(`Expression`~Health.Status, data=metabolism.genes_clean)

t.test(Expression ~ Health.Status, data = metabolism.genes_clean)$p.value

library(infer)
results <- metabolism.genes_clean %>% 
  specify(Expression ~ Health.Status) %>% 
  mutate(Expression = as.numeric(Expression)) %>% 
  mutate(Health.Status = as.factor(Health.Status)) %>% 
  specify(Expression ~ Health.Status) %>% 
  hypothesize(null = 'independence') %>% 
  generate(reps = 1000, type = 'permute') %>% 
  calculate("diff in means", order = c("Healthy", "Disease"))

obs <- metabolism.genes_clean %>% 
  mutate(Expression = as.numeric(Expression)) %>% 
  mutate(Health.Status = as.factor(Health.Status)) %>% 
  specify(Expression ~ Health.Status) %>% 
  # hypothesize(null = 'independence') %>% 
  calculate("diff in means", order = c("Healthy", "Disease"))

pval <- results %>% get_p_value(obs_stat = obs, direction='two_sided')
pval

results %>% 
  visualize() +
  shade_p_value(obs_stat=obs, direction='two_sided')
```

##Data Visualization
```{r}
#Graphically represent data
#We will be using the ggplot package. A Manhattan plot would allow better visual representation of gene expression. 

(metabolism.genes_graph <- metabolism.genes_clean %>% 
 ggplot()+
  geom_bar(mapping = aes(x=Gene.Symbol, y=Expression, fill=Health.Status), stat='identity', position='dodge')
)
```


**References**
  
Turner, S.D. (2014) **qqman: an R package for visualizing GWAS results using Q-Q and manhattan plots. biorXiv** DOI: 10.1101/005165 

Heat map generated using Morpheus from the Broad Institute

